pkgs/development/interpreters/jruby/default.nix

```diff
@@ -1,67 +1,106 @@
-{ lib, stdenv, callPackage, fetchurl, makeWrapper, jre }:
+{ lib, stdenv, callPackage, fetchFromGitHub, makeWrapper, jre, maven, ant, jrubyVersion ? "9.2.16.0" }:
 
 let
-# The version number here is whatever is reported by the RUBY_VERSION string
-rubyVersion = callPackage ../ruby/ruby-version.nix {} "2" "5" "7" "";
-jruby = stdenv.mkDerivation rec {
-  pname = "jruby";
-
-  version = "9.2.16.0";
+  # The version number here is whatever is reported by the RUBY_VERSION string
+  rubyVersion = callPackage ../ruby/ruby-version.nix { } "2" "5" "7" "";
+  jruby-src = fetchFromGitHub {
+    owner = "jruby";
+    repo = "jruby";
+    rev = "${jrubyVersion}";
+    sha256 = "14zs1xr9rb7ydna7kfr2xvsvgwpm8lr5ngmb8dqgdv1hq687h6wz";
+  };
+  # Nixpkgs works by disallowing Internet access in order to enforce
+  # reproducibility. One way around that however is if your build is already
+  # binary reproducible, which you can toggle by specifying outputHash.
+  # We create a maven repository (~/.m2) for use in the build further down.
+  maven-repository = stdenv.mkDerivation {
+    name = "jruby-maven-repository";
+    buildInputs = [ jre maven ];
+    src = jruby-src;
+    buildPhase = ''
+      mkdir $out
+      while mvn -Pdist -Dmaven.repo.local=$out -Dmaven.wagon.rto=5000; [ $? = 1 ]; do
+        echo "timeout, restart maven to continue downloading"
+      done
+    '';
+    # keep only *.{pom,jar,sha1,nbm} and delete all ephemeral files with lastModified timestamps inside
+    installPhase = ''
+      find $out -type f \
+        -name \*.lastUpdated -or \
+        -name resolver-status.properties -or \
+        -name _remote.repositories \
+        -delete
+    '';
+    # don't do any fixup
+    dontFixup = true;
 
-  src = fetchurl {
-    url = "https://s3.amazonaws.com/jruby.org/downloads/${version}/jruby-bin-${version}.tar.gz";
-    sha256 = "sha256-WuJ/FJ9z8/6k80NZy7dzwl2dmH5yte3snouTlXmX6zA=";
+    outputHashAlgo = "sha256";
+    outputHashMode = "recursive";
+    outputHash = "0prz7kzg1z4c0xakdwwh52gciwii35lbhrb5wsdw9p4wykllnw1x";
   };
+  jruby = stdenv.mkDerivation rec {
+    pname = "jruby";
 
-  nativeBuildInputs = [ makeWrapper ];
+    version = jrubyVersion;
 
-  installPhase = ''
-     mkdir -pv $out/docs
-     mv * $out
-     rm $out/bin/*.{bat,dll,exe,sh}
-     mv $out/COPYING $out/LICENSE* $out/docs
+    src = jruby-src;
 
-     for i in $out/bin/jruby{,.bash}; do
-       wrapProgram $i \
-         --set JAVA_HOME ${jre.home}
-     done
+    buildInputs = [ makeWrapper maven ant jre ];
 
-     ln -s $out/bin/jruby $out/bin/ruby
+    buildPhase = ''
+      echo "Using repository ${maven-repository}"
+      # We make sure to avoid installation since the maven repository is read-only now
+      mvn -Dmaven.install.skip=true -Pdist --offline -Dmaven.repo.local=${maven-repository}
+      mkdir ./dist
+      tar -xzvf ./maven/jruby-dist/target/jruby-dist-*-bin.tar.gz --strip-components=1 -C ./dist
+      cd ./dist
+    '';
```

Comment:
Kind of a bit of a waste to build the full tar distribution but according to JRuby's maintainer @headius, there's no interim accessible state as of now.